v1/topdown/query.go

```diff
@@ -544,6 +559,13 @@ func (q *Query) Iter(ctx context.Context, iter func(QueryResult) error) error {
 		vc = NewVirtualCache()
 	}
 
+	var bc BaseCache
+	if q.baseCache != nil {
+		bc = q.baseCache
+	} else {
+		bc = newBaseCache()
+	}
```

Comment:
Given that usage, I wonder if we're really adding what's needed: the old BC was instantiated here, and the new option only gives you a single instance.

What wouldn't surprise me would be this
```go
func (q *Query) WithBaseCache(bc func() BaseCache) *Query {
	q.baseCache = bc
	return q
}
```

and
```suggestion
	if q.baseCache != nil {
		bc = q.baseCache()
	} else {
		bc = newBaseCache()
	}
```

Reply:
Not sure I follow. We're just leaving the responsibility of instantiation to the caller (if they provide anything, that is). You can provide a single instance across evals if you want (like we do) and you can also instantiate a new cache per call if that's your preference. The virtual cache is instantiated the same way, so this was mostly just following that pattern... but I have no objections to changing this if I just understood the benefit.

Reply:
I misremembered what Iter() does. There's only one instance of BC ever. It's fine! Thanks for bearing with me.